text/001x-battery_improvements.md

```diff
@@ -0,0 +1,178 @@
+- a new (optional) `BATTERY_VOLTAGES` message that can be scaled to as many cells as needed, and which reports them as faults.
+
+  
+# Motivation
+
+The motivation is to:
+- reduce the memory required for battery reporting ([anecdotal evidence](https://github.com/ArduPilot/mavlink/pull/233#issuecomment-976197179) indicates cases with 15% of bandwidth on a default RF channel).
+- Provide a cleaner message and design for both implementers and consumers.
+
+The [BATTERY_STATUS](https://mavlink.io/en/messages/common.html#BATTERY_STATUS) has two arrays that can be used to report individual cell voltages (up to 14), or a single cumulative voltage, and which cannot be truncated.
+The vast majority of consumers are only interested in the total battery voltage, and either sum the battery cells or get the cumulative voltage.
+By separating the cell voltage reporting into a separate message the new battery status can be much smaller, and the cell voltages need only be sent if the consumer is actually interested.
+
+# Detailed Design
+
+There are three parts to the design:
+1. A more efficient status message
+2. A scalable battery voltage message.
+3. Mechanisms that allow the new messages to seamlessly coexist with the old one along with eventual deprecation of the older message.
+
+## BATTERY_STATUS_V2
+
+The proposed message is:
+
+```xml
+    <message id="???" name="BATTERY_STATUS_V2">
+      <description>Battery dynamic information. This should be streamed (nominally at 1Hz). Static battery information is sent in SMART_BATTERY_INFO.</description>
+      <field type="uint8_t" name="id" instance="true">Battery ID</field>
+      <field type="int16_t" name="temperature" units="cdegC" invalid="INT16_MAX">Temperature of the battery. INT16_MAX for unknown temperature.</field>
+      <field type="uint32_t" name="voltage" units="mV" invalid="[UINT32_MAX]">Battery voltage (total).</field>
+      <field type="int16_t" name="current" units="cA" invalid="UINT16_MAX">Battery current (through all cells/loads). Positive if discharging, negative if charging. UINT16_MAX: field not provided.</field>
+      <field type="int32_t" name="current_consumed" units="mAh" invalid="-1">Consumed charge, -1: Current consumption estimate not provided.</field>
+      <field type="int8_t" name="percent_remaining" units="%" invalid="-1">Remaining battery energy. Values: [0-100], -1: Remaining battery energy is not provided.</field>
+      <field type="uint32_t" name="time_remaining" units="s" invalid="UINT32_MAX">Remaining battery time (estimated), UINT32_MAX: Remaining battery time estimate not provided.</field>
+      <field type="uint32_t" name="fault_bitmask" display="bitmask" enum="MAV_BATTERY_FAULT">Fault/health/ready-to-use indications.</field>
+    </message>
+```
+
+With `MAV_BATTERY_FAULT` having the following **additions** (from `MAV_BATTERY_CHARGE_STATE` and `MAV_BATTERY_MODE`):
```

Comment:
I made a comment above about it but my opinion is we start with a clean slate and define a new enum BATTERY_STATUS_V2_STATUS_FLAGS

Reply:
@dakejahl I'm happy to be lead by you, but can you provide your proposed set? I'm completely ignorant, so I'll be floating both the set and the dropped things past various experts. The think about enums though is that they are easy to add afterwards, so a smaller set is fine by me.

Reply:
The set of flags in the DroneCAN PR are a good starting point. I am trying to think of any more we might want. The SAFETY_FAULT flag is used to indicate that one of the other fault conditions is a "safety fault", for BMS ASICs this would correspond to disabling the output FETs if protections are enabled (which is also indicated by a flag).
https://github.com/dronecan/DSDL/pull/7/files#diff-84419f8d2b61992ae9d50e2d09029c9cec37365ca16c6a29edae815177fc0c68R12-R26
```
uint16 STATUS_FLAG_IN_USE               = 1
uint16 STATUS_FLAG_CHARGING             = 2
uint16 STATUS_FLAG_OVER_TEMP            = 4
uint16 STATUS_FLAG_UNDER_TEMP           = 8
uint16 STATUS_FLAG_OVER_VOLT            = 16
uint16 STATUS_FLAG_UNDER_VOLT           = 32
uint16 STATUS_FLAG_OVER_CURRENT         = 64
uint16 STATUS_FLAG_SHORT_CIRCUIT        = 128
uint16 STATUS_FLAG_SAFETY_FAULT         = 256
uint16 STATUS_FLAG_CELL_IMBALANCE       = 512
uint16 STATUS_FLAG_CELL_BALANCING       = 1024
uint16 STATUS_FLAG_PROTECTIONS_ENABLED  = 2048
uint16 STATUS_FLAG_REQUIRES_SERVICE     = 4096
uint16 STATUS_FLAG_BAD_BATTERY          = 8192
uint16 STATUS_FLAG_ERROR_UNKNOWN        = 32768
```

Reply:
Also I am going to bump from uint16 to uint32 to this can be further extended

Reply:
@dakejahl Thanks. So I'm more than happy to _propose_ starting again with `MAV_BATTERY_STATUS` (rather than `MAV_BATTERY_FAULT`) but we'll see how that goes. It makes sense to me because now it is a general status and not just a fault. 

In MAVLink adding enum values that "belong" is a zero cost thing to do. Is it the same in UAVCAN? If so, if there are any of these things that are speculative then we might be better off not including them? 

So, starting with the MAVLink set, below are the ones that don't have an obvious mapping between sets (to me). I need to justify the changes, most likely by saying they aren't needed because:
- won't ever be reported because can't be detected
- not useful
- don't make sense
- -can't be acted on etc.

These are battery faults - good/bad/ remove/keep/ why? If replace - why?

Value | Field Name | Description
--- | --- | ---
1 | MAV_BATTERY_FAULT_DEEP_DISCHARGE | Battery has deep discharged.
2 | MAV_BATTERY_FAULT_SPIKES | Voltage spikes.
4 | MAV_BATTERY_FAULT_CELL_FAIL | One or more cells have failed. 

These are "battery is not compatible" flags. I.e.  incompatibility driver detects that the battery shouldn't work properly even though the battery itself is behaving correctly. I think 64 and 256 might be variations on a theme.

Value | Field Name | Description
--- | --- | ---
64 | MAV_BATTERY_FAULT_INCOMPATIBLE_VOLTAGE | Vehicle voltage is not compatible with this battery (batteries on same power rail should have similar voltage).
128 | MAV_BATTERY_FAULT_INCOMPATIBLE_FIRMWARE | Battery firmware is not compatible with current autopilot firmware.
256 | MAV BATTERY_FAULT_INCOMPATIBLE_CELLS_CONFIGURATION | Battery is not compatible due to cell configuration (e.g. 5s1p when vehicle requires 6s).

These are states, like charging

Value | Field Name | Description
--- | --- | ---
1024 | BATTERY_FAULT_AUTO_DISCHARGING  | Battery is auto discharging (towards storage level). Not ready to use.
2056 | BATTERY_FAULT_HOT_SWAP | Battery in hot-swap mode (current limited to prevent spikes that might damage sensitive electrical circuits). Not ready to use.

Thoughts?

Reply:
Then, looking at the UAVCAN set, below are the ones without a direct mapping. I assume these are all things that might really occur and would be _useful to report_ (i.e. if it is a bad battery, presumably you might want to know it is bad due to a short circuit?). 

The missing thing for me is docs - I don't know much about batteries, so need explanation of what they do in order to know if this is clear. I've put comments/questions below for each. 

- uint32 STATUS_FLAG_IN_USE               = 1
  - This is just to indicate that the battery provides status reporting right? So this would always be set to 1 in most smart batteries?
- uint32 STATUS_FLAG_OVER_VOLT            = 16
  - Is this that the battery is higher voltage that spec (and if so what kind of voltage level would trigger this - 10%?)
  - Or is this battery voltage is higher rated than allowed on bus - e.g. equivalent to over-battery case for MAV_BATTERY_FAULT_INCOMPATIBLE_VOLTAGE
  - Is this a critical fault? I.e. would it also trigger STATUS_FLAG_SAFETY_FAULT and STATUS_FLAG_REQUIRES_SERVICE?
- uint32 STATUS_FLAG_UNDER_VOLT           = 32
  - As above?
- uint32 STATUS_FLAG_SHORT_CIRCUIT        = 128
  - What other flags does this set - i.e. STATUS_FLAG_REQUIRES_SERVICE, bad battery? Is it throw away or service, or don't fly. 
- uint32 STATUS_FLAG_SAFETY_FAULT         = 256 
  - Is this "dont' fly" or "potential issue"? 
  - Essentially, where does it sit compared to STATUS_FLAG_BAD_BATTERY, STATUS_FLAG_REQUIRES_SERVICE and STATUS_FLAG_BAD_BATTERY. My assumption is probably critical, but maybe you can service the battery or throw it away, depending on the actual fault.
- uint32 STATUS_FLAG_CELL_IMBALANCE       = 512
  - I think this means that the cells have different voltages - is that right? What does a user do in this case? Is it a safety issue?
- uint32 STATUS_FLAG_CELL_BALANCING       = 1024 
  - I assume this is some kind of state like "Charging"? Can the battery be used? What other things get triggered.
- uint32 STATUS_FLAG_PROTECTIONS_ENABLED  = 2048 
  - What does this mean?
- uint32 STATUS_FLAG_REQUIRES_SERVICE     = 4096
  - As above, can you fly with this? What other flags might be set?  
- uint32 STATUS_FLAG_BAD_BATTERY          = 8192
  - Does this mean throw away? What else triggers this? 
- uint32 STATUS_FLAG_ERROR_UNKNOWN        = 4294967295
  - I assume this is something a battery can use if it wants to report an error that isn't one of the above.?

Reply:
@dakejahl OK, for the "battery is not compatible" flags, I asked much the same questions. The answers are here: https://github.com/mavlink/mavlink/pull/1577#discussion_r572404514

@cmic0 We're making a common set of flags for batteries with UAVCAN (if at all possible). Can you please respond to the comments below about the incompatible firmware, voltage, and cell configuration. In summary "bad things will happen with the battery before software gets to tell you there is a problem".


>> MAV_BATTERY_FAULT_INCOMPATIBLE_FIRMWARE

>  remove, not useful. How is the autopilot software supposed to determine this?

>> MAV_BATTERY_FAULT_INCOMPATIBLE_VOLTAGE

> remove, something bad is going to happen electrically before software gets a chance to report this. In general if you plug the wrong voltage power supply into a device it's not going to work.

>> MAV_BATTERY_FAULT_INCOMPATIBLE_CELLS_CONFIGURATION 

> remove, same logic as the first one, if the voltage is wrong bad things will happen electrically

Reply:
If i got the DroneCan flag correctly, `BATTERY_FAULT_HOT_SWAP` is actually the same as `STATUS_FLAG_PROTECTIONS_ENABLED`, hot swap was just another word to indicate that the battery protections to limit the current output are enabled, and defining them as "states" rather than "faults" clarifies things by a lot.

> MAV_BATTERY_FAULT_INCOMPATIBLE_FIRMWARE
In most of the implementations the driver on the autopilot side can have the knowledge about which minimum firmware version is required, when this flag is set means that the battery needs to be updated. I think this is a general condition worth to keep.

> MAV_BATTERY_FAULT_INCOMPATIBLE_VOLTAGE
In a multi-battery setup, you want to avoid that batteries with different charge levels are used for the flight. E.g if the delta between two batteries is > 2V the autopilot should flag it to the GCS.

> BATTERY_FAULT_INCOMPATIBLE_CELLS_CONFIGURATION
This flag is meant to catch the cases where the user tries to use a wrong battery on the vehicle (e.g a 5S instead of a 6S). In this circumstances the GCS should notify the user to correct the setup prior flying, does it make sense?

Reply:
>In most of the implementations the driver on the autopilot side can have the knowledge about which minimum firmware version is required

For a battery with a "driver" this could be the case. For a uavcan battery this is not the case. 

>In a multi-battery setup, you want to avoid that batteries with different charge levels are used for the flight. E.g if the delta between two batteries is > 2V the autopilot should flag it to the GCS.

>This flag is meant to catch the cases where the user tries to use a wrong battery on the vehicle (e.g a 5S instead of a 6S). In this circumstances the GCS should notify the user to correct the setup prior flying, does it make sense?

If they are in series, sure you can indicate and maybe this is a good thing. If they are in parallel the higher voltage battery is going to sink current into the lower voltage battery and bad things will happen.

Since these are just flags I'm fine with leaving them in.

Reply:
>In most of the implementations the driver on the autopilot side can have the knowledge about which minimum firmware version is required

For a battery with a "driver" this could be the case. For a uavcan battery this is not the case. 

>In a multi-battery setup, you want to avoid that batteries with different charge levels are used for the flight. E.g if the delta between two batteries is > 2V the autopilot should flag it to the GCS.

>This flag is meant to catch the cases where the user tries to use a wrong battery on the vehicle (e.g a 5S instead of a 6S). In this circumstances the GCS should notify the user to correct the setup prior flying, does it make sense?

If they are in series, sure you can indicate and maybe this is a good thing. If they are in parallel the higher voltage battery is going to sink current into the lower voltage battery and bad things will happen.

Since these are just flags I'm fine with leaving them in.

Reply:
> If i got the DroneCan flag correctly, BATTERY_FAULT_HOT_SWAP is actually the same as STATUS_FLAG_PROTECTIONS_ENABLED, 

@cmic0 They might be the same but there might be a subtle difference.

`STATUS_FLAG_PROTECTIONS_ENABLED` indicates that a safety system is _enabled_ but not necessarily "engaged". In other words, if this is true it is monitoring while supplying power as normal. IFF it detects a problem, e.g. a short circuit, it might trigger a STATUS_FLAG_PROTECTIONS_FAULT and turns off power flow.

Is that the same as BATTERY_FAULT_HOT_SWAP? (my read of your comment above might be that in HOT swap you have explicitly engaged that same state as `STATUS_FLAG_PROTECTIONS_FAULT` - i.e. no power, not that you are monitoring for issues. 

@dakejahl @cmic0 Irrespective of what "hot swap" does, is there a case for a state "like" STATUS_FLAG_PROTECTIONS_ENGAGED where you say I'm not monitoring and I don't have a fault, but things are risky so I want to turn off the outputs?

Reply:
For mavlink we can keep `BATTERY_FAULT_INCOMPATIBLE_CELLS_CONFIGURATION`. A system can't necessarily report this, but it is IMO fine to have status if it can be reported by some means. We should probably document that not all "faults" can be reported by all systems.

>> `BATTERY_FAULT_INCOMPATIBLE_CELLS_CONFIGURATION`
>>
> > This flag is meant to catch the cases where the user tries to use a wrong battery on the vehicle (e.g a 5S instead of a 6S). In this circumstances the GCS should notify the user to correct the setup prior flying, does it make sense?
> 
> If they are in series, sure you can indicate and maybe this is a good thing. If they are in parallel the higher voltage battery is going to sink current into the lower voltage battery and bad things will happen.
> 
> Since these are just flags I'm fine with leaving them in.

I'm fine with leaving it in, but I would like you @cmic0 to please clarify the parallel case. Because I suspect they might be in parallel, and I still wonder how this is useful. Or is it that the bad things are happening, and you think that there is still time to report the problem before it all falls over?